src/server/scripts/Outland/CoilfangReservoir/TheSlavePens/boss_ahune.cpp

```diff
@@ -0,0 +1,616 @@
+                for (int counter = 1; counter < 7; counter++)
+                {
+                    if (counter < 4)
+                        me->SummonCreature(NPC_EARTHEN_RING_TOTEM, SummonPositions[counter], TEMPSUMMON_MANUAL_DESPAWN);
+                    else
+                        me->SummonCreature(NPC_EARTHEN_RING_FLAMECALLER, SummonPositions[counter], TEMPSUMMON_MANUAL_DESPAWN);
+                }
+            }
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            if (say_timer <= diff)
+            {
+                if (Player* player = SelectRandomPlayer(100.0f, false))
+                {
+                    if (count == 0)
+                        sCreatureTextMgr->SendChat(me, SAY_PLAYER_TEXT_1, NULL, CHAT_MSG_SAY, LANG_UNIVERSAL, TEXT_RANGE_NORMAL, 0, TEAM_OTHER, false, player);
+                    if (count == 1)
+                        sCreatureTextMgr->SendChat(me, SAY_PLAYER_TEXT_2, NULL, CHAT_MSG_SAY, LANG_UNIVERSAL, TEXT_RANGE_NORMAL, 0, TEAM_OTHER, false, player);
+                    if (count == 2)
+                        sCreatureTextMgr->SendChat(me, SAY_PLAYER_TEXT_3, NULL, CHAT_MSG_SAY, LANG_UNIVERSAL, TEXT_RANGE_NORMAL, 0, TEAM_OTHER, false, player);
+                    if (count == 3)
+                        TempSummon* ahune = me->SummonCreature(NPC_AHUNE, SummonPositions[0], TEMPSUMMON_DEAD_DESPAWN);
+                    ++count;
+                    say_timer = 3000;
+                }
+            }
+            else
+                say_timer -= diff;
+        }
+
+        Player* SelectRandomPlayer(float range = 0.0f, bool checkLoS = true)
+        {
+            Map* map = me->GetMap();
+            if (!map->IsDungeon())
+                return NULL;
+
+            Map::PlayerList const &PlayerList = map->GetPlayers();
+            if (PlayerList.isEmpty())
+                return NULL;
+
+            std::list<Player*> temp;
+            for (Map::PlayerList::const_iterator i = PlayerList.begin(); i != PlayerList.end(); ++i)
+                if ((me->IsWithinLOSInMap(i->GetSource()) || !checkLoS) && me->GetVictim() != i->GetSource() &&
+                    me->IsWithinDistInMap(i->GetSource(), range) && i->GetSource()->IsAlive())
+                    temp.push_back(i->GetSource());
+
+            if (!temp.empty())
+                Trinity::Containers::SelectRandomContainerElement(temp);
+        }
+    };
+
+    CreatureAI* GetAI(Creature* creature) const override
+    {
+        return new npc_ahune_bunnyAI(creature);
+    }
+};
+
+class npc_ahune_ice_spear : public CreatureScript
+{
+public:
+    npc_ahune_ice_spear() : CreatureScript("npc_ice_spear_bunny") { }
+
+    struct npc_ahune_ice_spearAI : public ScriptedAI
+    {
+        npc_ahune_ice_spearAI(Creature* creature) : ScriptedAI(creature)
+        {
+            _spikeTimer = 2500;
+            _spiked = false;
+            DoCast(me, SPELL_ICE_SPEAR_VISUAL);
```

Comment:
I'm not sure if casting spell inside the constructor is a good idea... Perhaps you can move these spells to `auras` in `creature_template_addon`? Check sniffs.